crates/extensions/c8y_mapper_ext/src/operations/mod.rs

```diff
@@ -13,47 +13,270 @@
+    pub tedge_http_host: Arc<str>,
+    pub tmp_dir: Arc<Utf8Path>,
+    pub mqtt_schema: MqttSchema,
+    pub c8y_prefix: tedge_config::TopicPrefix,
+    pub software_management_api: SoftwareManagementApiFlag,
+    pub command_id: IdGenerator,
+
+    pub http_proxy: C8YHttpProxy,
+    pub c8y_endpoint: C8yEndPoint,
+    pub auth_proxy: ProxyUrlGenerator,
+
+    pub downloader: ClientMessageBox<IdDownloadRequest, IdDownloadResult>,
+    pub uploader: ClientMessageBox<IdUploadRequest, IdUploadResult>,
+    pub mqtt_publisher: LoggingSender<MqttMessage>,
+
+    pub running_operations: Arc<Mutex<HashMap<Arc<str>, RunningOperation>>>,
+}
+
+impl OperationHandler {
+    /// Handles an MQTT command id message.
+    ///
+    /// All MQTT messages with a topic that contains an operation id, e.g.
+    /// `te/device/child001///cmd/software_list/c8y-2023-09-25T14:34:00` need to be passed here for
+    /// operations to be processed. Messages not related to operations are ignored.
+    ///
+    /// `entity` needs to be the same entity as in `message`.
+    ///
+    /// When a message with a new id is handled, a task will be spawned that processes this
+    /// operation. For the operation to be completed, subsequent messages with the same command id
+    /// need to be handled as well.
+    ///
+    /// When an operation terminates (successfully or unsuccessfully), an MQTT operation clearing
+    /// message will be published automatically, which does not need to be handled.
+    pub async fn handle(self: &Arc<Self>, entity: EntityTarget, message: MqttMessage) {
+        let Ok((_, channel)) = self.mqtt_schema.entity_channel_of(&message.topic) else {
+            return;
+        };
+
+        let Channel::Command { operation, cmd_id } = channel else {
+            return;
+        };
+
+        let message = OperationMessage {
+            operation,
+            entity,
+            cmd_id: cmd_id.into(),
+            message,
+        };
+
+        let topic: Arc<str> = message.message.topic.name.clone().into();
+        let running_operation = {
+            let mut lock = self.running_operations.lock().unwrap();
+            let op = lock.get(&topic);
+
+            if let Some(running_operation) = op {
+                // task already terminated
+                if running_operation.tx.send(message).is_err() {
+                    let running_operation = lock.remove(&topic).unwrap();
+                    Some(running_operation)
+                } else {
+                    None
+                }
+            } else {
+                let handler = self.clone();
+                let running_operation = RunningOperation::spawn(message, handler);
+
+                lock.insert(topic, running_operation);
+                None
+            }
+        };
+
+        if let Some(running_operation) = running_operation {
+            let join_result = running_operation.handle.await;
+            error!("handle: {join_result:?}");
+        }
+    }
```

Comment:
The name `running_operation` is confusing. Yes, there was a running operation, but this is operation is no more consuming messages (`running_operation.tx.send(message).is_err()`).

```suggestion
        if let Some(terminated_operation) = terminated_operation {
            let join_result = terminated_operation.handle.await;
            error!("handle: {join_result:?}");
        }
    }
```

By the way, will the entry in `self.running_operations` for a topic be cleared, if no more message is received for that topic?

Reply:
> By the way, will the entry in self.running_operations for a topic be cleared, if no more message is received for that topic?

No. I realized that currently, after the operation has terminated, the converter has to call `handle()` with one additional message, and in practice the only message that can be used is the topic clearing message, because it's published by the operation on MQTT broker, and the broker echoes clearing message back to us because we're subscribed, at which point converter can call `handle()` and only then is this task removed from the map.

So the current design is not unlike polling, where the caller (converter in this case) needs to call `.handle()` for every operation-related message for the operation to be successfully handled and freed.

This has some problems, the biggest being that we can't remove the operation immediately after it terminates, and have to rely on MQTT broker to echo back the clearing message, which shouldn't be necessary, but I think this could be addressed in the following PR.

Reply:
will be done in a follow-up: https://github.com/thin-edge/thin-edge.io/issues/2880#issuecomment-2223213567